src/configurationProvider.ts

```diff
@@ -0,0 +1,101 @@
+import * as serverUtils from './omnisharp/utils';
+import { AssetGenerator, addTasksJsonIfNecessary, createLaunchConfiguration, createAttachConfiguration, containsDotNetCoreProjects, createWebLaunchConfiguration } from './assets';
+import { isSubfolderOf } from './common';
+
+export class CSharpConfigurationProvider implements vscode.DebugConfigurationProvider {
+    private server: OmniSharpServer;
+
+    public constructor(server: OmniSharpServer) {
+        this.server = server;
+    }
+
+    /**
+     * TODO: Remove function when https://github.com/OmniSharp/omnisharp-roslyn/issues/909 is resolved.
+     * 
+     * Note: serverUtils.requestWorkspaceInformation only retrieves one folder for multi-root workspaces. Therefore, generator will be incorrect for all folders
+     * except the first in a workspace. Currently, this only works if the requested folder is the same as the server's solution path or folder.
+     */
+    private checkWorkspaceInformationMatchesWorkspaceFolder(folder: vscode.WorkspaceFolder | undefined): boolean {
+        const solutionPathOrFolder: string = this.server.getSolutionPathOrFolder();
+
+        // Make sure folder, folder.uri, and solutionPathOrFolder are defined.
+        if (!folder || !folder.uri || !solutionPathOrFolder)
+        {
+            return false;
+        }
+
+        let serverFolder = solutionPathOrFolder;
+        // If its a .sln file, get the folder of the solution.
+        if (fs.lstatSync(solutionPathOrFolder).isFile())
+        {
+            serverFolder = path.dirname(solutionPathOrFolder);
+        }
+
+        // Get absolute paths of current folder and server folder.
+        const currentFolder = path.resolve(folder.uri.fsPath);
+        serverFolder = path.resolve(serverFolder);
+
+        return currentFolder && folder.uri && isSubfolderOf(serverFolder, currentFolder);
+    }
+
+    /**
+	 * Returns a list of initial debug configurations based on contextual information, e.g. package.json or folder.
+	 */
+    provideDebugConfigurations(folder: vscode.WorkspaceFolder | undefined, token?: vscode.CancellationToken): vscode.ProviderResult<vscode.DebugConfiguration[]> {
+        return serverUtils.requestWorkspaceInformation(this.server).then(info => {
+            const generator = new AssetGenerator(info);
+            if (this.checkWorkspaceInformationMatchesWorkspaceFolder(folder) && containsDotNetCoreProjects(info)){
+                const dotVscodeFolder: string = path.join(folder.uri.fsPath, '.vscode');
+                const tasksJsonPath: string = path.join(dotVscodeFolder, 'tasks.json');
+                
+                // Make sure .vscode folder exists, addTasksJsonIfNecessary will fail to create tasks.json if the folder does not exist. 
+                fs.ensureDirSync(dotVscodeFolder);
+    
+                // if the file does not exist, addTasksJson
+                const addTasksJson: boolean = !fs.existsSync(tasksJsonPath);
```

Comment:
I'm preparing a PR to take advantage of the taskProvider API. We'll want to keep these two implementations in-sync. I'm also adding quite a bit of test infrastructure that we can use to ensure that they are in sync.